service/worker/scheduler/workflow.go

```diff
@@ -334,6 +361,38 @@ func (s *scheduler) processPatch(patch *schedpb.SchedulePatch) {
 	}
 }
 
+func (s *scheduler) getNextTime(after time.Time, stop time.Time) getNextTimeResult {
+
+	// ignore start times that come before the 'after' parameter
+	for len(s.nextTimeResultCache) != 0 && (s.nextTimeResultCache[0].Next.Before(after) || s.nextTimeResultCache[0].Next.Equal(after)) {
+		s.nextTimeResultCache = s.nextTimeResultCache[1:]
+	}
+
+	// if we exhausted all cache elements, fetch more schedules
+	if len(s.nextTimeResultCache) == 0 {
+		// Run this logic in a SideEffect so that we can fix bugs there without breaking
+		// existing schedule workflows.
+		panicIfErr(workflow.SideEffect(s.ctx, func(ctx workflow.Context) interface{} {
+			var results []getNextTimeResult
+			for len(results) < maxNextTimeResultCacheSize {
+				next := s.cspec.getNextTime(after)
+				after = next.Next
+				results = append(results, next)
+				// if there is no matching time, stop fetching
+				if after.IsZero() {
+					break
+				}
+			}
+			return results
+		}).Get(&s.nextTimeResultCache))
+	}
+	next := s.nextTimeResultCache[0]
+	if next.Next.Before(stop) || next.Next.Equal(stop) {
```

Comment:
I see, valid point; the assumption was this is going to be called sequentially. I guess the first approach is simpler, let me implement it and see how it turns out.

Reply:
`stop` was solving another problem; not discarding the last time result returned from `getNextTime`. But I figured there is no need to discard anything from the cache. By using a map instead of array (`after -> next`), it becomes simpler; if this is called sequentially, `after` is always matches what is passed in for cached values, and if `after` does not match any existing items, it means we need to fetch again. If it is not called sequentially,  discard the whole cache and fetch again.

Reply:
I guess you were stating the same thing